keps/sig-storage/1979-object-storage-support/README.md

```diff
@@ -0,0 +1,730 @@
+1. `protocol.name`: (required) specifies the desired protocol.  One of {“s3”, “gs”, or “azureBlob”}.
+1. `protocol.version`: (optional) specifies the desired version of the `protocol`. For "s3", a value of "v2" or "v4" could be used. 
+1. `bucketPrefix`: (optional) prefix prepended to a generated new bucket name, eg. “yosemite-photos-". If `bucketInstanceName` is supplied then `bucketPrefix` is ignored because the request is for access to an existing bucket. If `bucketPrexix` is specified then it is added to the `Bucket` instance as a label.
+1. `bucketClassName`: (optional for greenfield, omitted for brownfield) name of the `BucketClass` used to provision this request. If omitted for a greenfield bucket request, a default bucket class matching the protocol, if available, is used. If the greenfield bucket class is missing or does not support the requested protocol, an error is logged and the request is retried (with exponential backoff). A `BucketClass` is necessary for greenfield requests since BCs support a list of allowed namespaces. A `BucketClass` is not needed for brownfield requests since the `Bucket` instance, created by the admin, also contains `allowedNamespaces`.
+1. `bucketInstanceName`: (optional) name of the cluster-wide `Bucket` instance. If blank, then COSI assumes this is a greenfield request and will fill in the name during the binding step. If defined by the user, then this names the `Bucket` instance created by the admin. There is no automation available to make this name known to the user.
+1. `bucketAvailable`: if true the bucket has been provisioned. If false then the bucket has not been provisioned and is unable to be accessed.
+
+#### Bucket
+
+A cluster-scoped custom resource representing the abstraction of a single backend bucket. A `Bucket` instance stores enough identifying information so that drivers can accurately target the backend object store (e.g. needed during a deletion process).  The relevant bucket class fields are copied to the `Bucket`. Additionally, data returned by the driver is copied to the `Bucket` by the sidecar.
+
+For greenfield, COSI creates the `Bucket` based on values in the `BucketRequest` and `BucketClass`. For brownfield, an admin manually creates the `Bucket`, filling in BucketClass fields, such as `allowedNamespaces`. COSI populates fields returned by the provisioner, and binds the `Bucket` to the `BucketAccess`.
+
+Since there is a 1:1 mapping between a BR and a `Bucket`, when multiple BRs request access to the same backend bucket, each separate `Bucket` points to the same backend bucket.
+
+```yaml
+apiVersion: cosi.io/v1alpha1
+kind: Bucket
+metadata:
+  name: [1]
+  labels: [2]
+    cosi.io/provisioner:
+    cosi.io/bucket-prefix:
+  finalizers:
+    - cosi.io/finalizer [3]
+spec:
+  provisioner: [4]
+  retentionPolicy: [5]
+  anonymousAccessMode: [6]
+  bucketClassName: [7]
+  bucketRequest: [8]
+  allowedNamespaces: [9]
+    - name:
+  protocol: [10]
```

Comment:
Similar to my comment on BucketRequest, we should consider making this a list of protocols to permit a object storage system to advertise that it supports more than one protocol and let the consumer pick which one it wants to use.

Reply:
See my reply above re. `BR.protocol`

Reply:
> Along the lines of intuitive and less complex, we are considering eliminating protocol from the BR. The reason why is that the BC reference in the BR is sufficient to know the protocol. The only value today in BR.protocol is that COSI can validate the BC's protocol matches the BR's. If validation fails then the BR is retried as is typical in k8s. An issue with this is the BC is immutable so the admin would need to recreate the BC with the correct protocol. Or, the BR needs to be deleted and recreated to use the right protocol. Both of these scenarios are unnecessary if BR.prtocol is removed.

Remember that BC may be deleted or changed (deleted and recreated with different values but same name), therefore Bucket should not depend on a reference to BC -- it has to be self sufficient even if the BC used to create it is deleted.

Reply:
We copy all BC fields to the `Bucket instance